core/trino-main/src/main/java/io/trino/operator/project/PageFieldsToInputParametersRewriter.java

```diff
@@ -42,77 +45,113 @@ private PageFieldsToInputParametersRewriter() {}
     public static Result rewritePageFieldsToInputParameters(RowExpression expression)
     {
         Visitor visitor = new Visitor();
-        RowExpression rewrittenProjection = expression.accept(visitor, null);
-        InputChannels inputChannels = new InputChannels(visitor.getInputChannels());
+        RowExpression rewrittenProjection = expression.accept(visitor, true);
+        InputChannels inputChannels = new InputChannels(visitor.getInputChannels(), visitor.getEagerlyLoadedChannels());
         return new Result(rewrittenProjection, inputChannels);
     }
 
     private static class Visitor
-            implements RowExpressionVisitor<RowExpression, Void>
+            implements RowExpressionVisitor<RowExpression, Boolean>
     {
         private final Map<Integer, Integer> fieldToParameter = new HashMap<>();
         private final List<Integer> inputChannels = new ArrayList<>();
+        private final Set<Integer> eagerlyLoadedChannels = new HashSet<>();
         private int nextParameter;
 
         public List<Integer> getInputChannels()
         {
             return ImmutableList.copyOf(inputChannels);
         }
 
+        public List<Integer> getEagerlyLoadedChannels()
+        {
+            return ImmutableList.copyOf(eagerlyLoadedChannels);
+        }
+
         @Override
-        public RowExpression visitInputReference(InputReferenceExpression reference, Void context)
+        public RowExpression visitInputReference(InputReferenceExpression reference, Boolean unconditionallyEvaluated)
         {
-            int parameter = getParameterForField(reference);
+            int parameter = getParameterForField(reference, unconditionallyEvaluated);
             return field(parameter, reference.getType());
         }
 
-        private Integer getParameterForField(InputReferenceExpression reference)
+        private Integer getParameterForField(InputReferenceExpression reference, boolean unconditionallyEvaluated)
         {
             return fieldToParameter.computeIfAbsent(reference.getField(), field -> {
                 inputChannels.add(field);
+                if (unconditionallyEvaluated) {
```

Comment:
I think this method can be called 2 times with reference for same field. If the latter call is with `unconditionallyEvaluated=true` (and former with `false`) then channel won't be loaded eagerly.
Can we have a test for that?

Reply:
yes, missed that, updated this now and added test case `COALESCE(0, bigint0) + bigint0`